src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,210 @@
+        /// <param name="semanticVersion">The document's semantic version associated with the passed-in tokens.</param>
+        /// <param name="range">The range associated with the passed-in tokens.</param>
+        /// <param name="tokens">The complete set of tokens for the passed-in range.</param>
+        internal void CacheTokens(
+            DocumentUri uri,
+            VersionStamp semanticVersion,
+            Range range,
+            int[] tokens)
+        {
+            if (!_cache.TryGetValue(uri, out var documentCache))
+            {
+                documentCache = new MemoryCache<VersionStamp, Dictionary<int, List<int>>>(sizeLimit: MaxSemanticVersionPerDoc);
+                _cache.Set(uri, documentCache);
+            }
+
+            if (!documentCache.TryGetValue(semanticVersion, out var lineToTokensDict))
+            {
+                lineToTokensDict = new Dictionary<int, List<int>>();
+                documentCache.Set(semanticVersion, lineToTokensDict);
+            }
+
+            var absoluteLine = 0;
+
+            // Cache the tokens associated with each line
+            for (var tokenIndex = 0; tokenIndex < tokens.Length; tokenIndex += 5)
+            {
+                absoluteLine += tokens[tokenIndex];
+
+                // Don't cache tokens outside of the given range
+                if (absoluteLine < range.Start.Line || absoluteLine > range.End.Line)
+                {
+                    continue;
+                }
+
+                if (lineToTokensDict.TryGetValue(absoluteLine, out _))
+                {
+                    // This line is already cached, so we can skip it
+                    continue;
+                }
+
+                var lineTokens = new List<int>();
+
+                // Cache the line's tokens
+                while (tokenIndex < tokens.Length)
+                {
+                    lineTokens.Add(tokens[tokenIndex]);
+                    lineTokens.Add(tokens[tokenIndex + 1]);
+                    lineTokens.Add(tokens[tokenIndex + 2]);
+                    lineTokens.Add(tokens[tokenIndex + 3]);
+                    lineTokens.Add(tokens[tokenIndex + 4]);
+
+                    if (tokens[tokenIndex + 5] != 0)
+                    {
+                        // We've hit the next line
+                        break;
+                    }
+
+                    tokenIndex += 5;
+                }
+
+                lineToTokensDict.Add(absoluteLine, lineTokens);
+            }
+
+            // Fill in the gaps for empty lines. When looking up tokens later, this helps us differentiate
+            // between empty lines vs. lines we don't have cached tokens for
+            for (var lineIndex = range.Start.Line; lineIndex < range.End.Line; lineIndex++)
+            {
+                if (!lineToTokensDict.TryGetValue(lineIndex, out _))
+                {
+                    lineToTokensDict.Add(lineIndex, new List<int>());
```

Comment:
If you're able to make the dictionary take an IReadOnlyList you can use array empty:
```suggestion
                    lineToTokensDict.Add(lineIndex, Array.Empty<int>());
```